### PR TITLE
Conduct experiment: TF-IDF + equal weight aggregation  + KNN cosine similarity filtering

### PR DESCRIPTION
**Pull request by katie-lamb**
Add a blocking experiment module with classes that handle adding in new embedding experiments. 

Implement KNN cosine similarity to filter the embedded tuples to get a set of likely candidate set tuples.

Start with implementing TF-IDF attribute embedding (n-gram or full word) + equal weight aggregation of attribute embeddings into tuple embeddings + KNN cosine similarity search.

Things to think about:

- Does it make sense that there are EIA utilities for records in the plant part list that aren't in any of the PUDL EIA utilities tables? They have valid `utility_id_eia` but no utility name. This is creating null values in the `utility_name` attribute
- Should the candidate set be created from only using the FERC vectors as query vectors ("the k best EIA matches for each FERC record") or also use the EIA vectors ("the k best EIA matches for each FERC record + the k best FERC record for each EIA record"). It seems like it should just be the FERC vectors as query vectors because there are so many extra EIA records that don't have a match on the FERC side. But this might exclude some potential matches.
  - Initial experiment with k = 5
     - 100% training data matches recall when EIA records are the query vectors
     - 95% training data match recall when FERC records are the query vectors
- Experiment with cosine similarity and other FAISS indexes, started with L2 (Euclidean distance).
- For now, I'm pinning the PUDL dependency back to `dev`, but if nightly builds start getting run on `main` then this should be pinned to `main`.

Null Values Notes:
- FAISS throws out null values when conducting the similarity search
- Ways to handle nulls (ordered roughly from easiest to hardest):
  - Imputation: fill with mean or median value
  - Encode as zeroes in the sparse representation
  - Weighted similarity: Weight the similarity calculation to down-weight the contribution of dimensions with missing values. Are null values concentrated in a few dimensions?
  - Interpolation: Try to estimate missing values based on non-missing values

To Do:

- [x] make a notebook that cleanly documents experiments and prints out metrics about the blocking set
- [x] vectorize the for loop that generates the query vectors for the similarity search
- [x] try using cosine similarity
- [x] block on report year
- [x] try using faster index search method
- [ ] add a function in the input module that makes training data at the true granularity

Would be nice to do:

- [ ] fill `capacity_mw` values with the median value from within a group of the same fuel type records. Right now it's filling with a global median
- [ ] run on all report years (just doing 2018 - 2020 right now)

Questions:
- Where should blocking results be stored? 
- How should we handle experiment tracking?
- Memory issues running on all years of data
- Should the training data be package data?
- What's up with that threading error?

### CODE REVIEW

**Comment by zaneselvans** [comment]
What are the records with NULL utility names like?  Are these real plants with lots of rich attribute information that are associated with utilities that we have little to no information about?  Or are they records where both the plant and the utility are kind of blank?

If the utilities have lots of information and just lack names, could it be that these are utilities with inconsistently reported names that got dumped by the harvesting / entity resolution process as too inconsistent?  I think there's a fallback utility name that we could be using instead that's always populated (from the `utilities_eia` table maybe?).  Or we could and probably should treat utility name as an annually varying field so that we retain that information in its original form for matching purposes (e.g. if a utility changes it's name over time, probably that will be reflected in multiple data sources, and having whatever year specific name it held stored in the data would be helpful).

File: .github/workflows/tox-pytest.yml [hunk 1]
```diff
@@ -59,7 +59,7 @@ jobs:
           conda run -n ferc1_eia_match tox
 
       - name: Upload test coverage report to CodeCov
-        uses: codecov/codecov-action@3.1.2
+        uses: codecov/codecov-action@v3.1.2
```
Comment (katie-lamb) [inline .github/workflows/tox-pytest.yml:62]: @zaneselvans any idea why this needs to be `v3.1.2` for the CI to pass but dependabot made it `3.1.2` in #50 ?

Comment (zaneselvans) [reply .github/workflows/tox-pytest.yml:62]: Hmm, dunno. That sounds buggy to me.

**Comment by zaneselvans** [comment]
@katie-lamb Should I be taking a look at this? Do we want to find a time to walk through it together before it fades from your memory?

**Comment by katie-lamb** [comment]
@zaneselvans 
**Summary of results for 2018 - 2020 data with TF-IDF vectorization and equal weight aggregation of attribute embeddings**

Results with L2 distance search (similar results for cosine similarity):
**k = 10**
- 96% of training match pairs are captured within the candidate set
- Candidate set is 16% the size of possible pairs with no blocking

**k = 5**
- 92% of training match pairs are captured within the candidate set
- Candidate set is 11% the size of possible pairs with no blocking

It takes 2-3 minutes to run this entire blocking step right now.

Most of the misses are because of nulls in FERC or EIA data. 

Possible next steps to improve results:
- Better imputation of missing values: filling in with median values for numeric and empty strings for string columns
- Include more matching columns. Right now I'm just using `plant_name`, `utility_name`, `fuel_type_code_pudl`, `installation_year`, `construction_year`, `capacity_mw`
- More complex vectorization and aggregation experiments.

It seems like results are good enough for now and I suggest that I keep moving forward with taking this candidate set and running it through `splink`.

File: src/ferc1_eia_match/candidate_set_creation.py [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+import pandas as pd
+import sqlalchemy as sa
+from scipy.sparse import hstack, issparse
+from sklearn.feature_extraction.text import TfidfVectorizer
+from sklearn.preprocessing import MinMaxScaler
+
+logger = logging.getLogger(__name__)
+
+
+class DataframeEmbedder:
```
Comment (katie-lamb) [inline src/ferc1_eia_match/candidate_set_creation.py:15]: This class takes the left and right dataframe as input, and creates an embedding matrix of the dataframes that are ready to be inputted into the similarity search.

The string columns (plant_name, utility_name, fuel_type_code_pudl) are vectorized with TF-IDF and null values are filled with the empty string. For fuel_type_code_pudl, a little bit of imputation is attempted, where if there is the FTCP in the plant_name, then the null value is filled in. E.g. "Miller Hydro"

The numeric columns (installation_year, construction_year, capacity_mw) are normalized between 0 and 1 and null values for filled in with the median value. If an installation_year is null but construction_year is not, then the installation_year is filled in with the construction_year value, and vice versa.

After vectorizing all the columns, the embeddings are stored in `col_embedding_dict` and then concatenated all together into one matrix using equal weight for each column.

File: src/ferc1_eia_match/candidate_set_creation.py [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+            self.left_embedding_attribute_dict
+        )
+        self.right_embedding_matrix = self.equal_weight_aggregate(
+            self.right_embedding_attribute_dict
+        )
+        if blocking_col != "":
+            self.set_col_blocks(blocking_col=blocking_col)
+
+
+class SimilaritySearcher:
```
Comment (katie-lamb) [inline src/ferc1_eia_match/candidate_set_creation.py:259]: This class conducts the similarity search, in this case finding k best EIA records for each FERC record. 

The query embedding matrix is the records that you want to find the k best of the other side for. In our case, FERC.
The menu embedding matrix is the records from which to search for the k best from. In our case, EIA.

I implemented raw L2 distance minimization search as well as maximizing cosine similarity search (same as taking the inner product of normalized vectors).

File: src/ferc1_eia_match/candidate_set_creation.py [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+                vectors for which to search for the k most similar matches from the
+                ``menu_embedding_matrix`` i.e. "for each query vector, find the top k most
+                similar vectors from the index of potential matches".
+            menu_embedding_matrix: The embedding matrix representing the potential match vectors
+                for the query vectors. These vector embeddings which will be searched over to find
+                vectors most similar or closest to the query vectors.
+            base_index (list): The base index for ``menu_embedding_matrix`` that will be
+                searched over for matches. Used to map back to the original vectors if the
+                embeddings were transformed in some way. Default is None and the search
+                will return the position of the vectors in the embedding matrix.
```
Comment (katie-lamb) [inline src/ferc1_eia_match/candidate_set_creation.py:285]: An example of how I use this `base_index` is in the blocking experiments notebook. Basically it helps to map from the outputted index within the report_year block back to the index of the full dataframe.

**Comment by katie-lamb** [comment]
I haven't set up infrastructure to compare results between experiments. Right now I just print out some stats on training set match retention and candidate set size in the blocking experiments notebook.

**Comment by katie-lamb** [comment]
@zaneselvans 

Here's the general flow of this PR:

- Data is preprocessed in `inputs.py` and made ready for the blocking step.
- The `blocking_experiments` notebook is what coordinates the blocking step right now.
- First, a `DataframeEmbedder` object vectorizes the left and right dataframes using TF-IDF for string columns (`plant_name`, `utility_name`, `fuel_type_code_pudl`) and scaling between 0 and 1 for numeric columns (`installation_year`, `construction_year`, `capacity_mw`). The vectorizations for each column are concatenated together with equal weighting to get one embedding matrix for each dataframe.
- Then, a `SimilaritySearcher` object performs either an L2 distance search or cosine similarity search (very similar results) to get the candidate set of k most similar EIA records for each FERC record. The indices output in this candidate set map back to the original EIA and FERC dataframes, and thus, we create a candidate set of records to be matched.
- A `block_num` is assigned for each FERC record and its k corresponding EIA match candidates. In the modeling step, we can block on this `block_num` (as well as `report_year`) to reduce the number of comparisons significantly. (I think this is the right thing to do).
- The `blocking_experiments` notebook prints out some stats on the performance of the blocking - how well does it capture training data matches, what is the reduction in number of comparisons.
- See [this comment](https://github.com/catalyst-cooperative/ccai-entity-matching/pull/51#issuecomment-1571009296) for a summary of the results.

File: src/ferc1_eia_match/candidate_set_creation.py (hunk 3 above)
Comment (zaneselvans) [reply src/ferc1_eia_match/candidate_set_creation.py:226]: In the original comment above you mentioned that there was 95% recall when the query vector was FERC but 100% recall when it was EIA, but we need for it to be FERC right? Since those are the "real" records, and the EIA side of things is all synthetic records?

**Comment by zaneselvans** [comment]
What does "% of comparisons with no blocking" mean?  Is that the fraction of records that are in the same block, and thus available to be considered as candidate matches? If so, those numbers are all tiny! Seems like the brute force annual blocking plus KNN is extremely effective, even if we're pretty generous with the value of K.

Why does it make sense to encode the categorical `fuel_type_code_pudl` column using TF-IDF rather than e.g. a one-hot encoder?

Are you surprised that the equal weighting works so well? What does that mean about the information we're comparing?

**Review by zaneselvans — COMMENTED** [review]
I still feel confused about how blocks of rows and the column embeddings are being stored / accessed. Maybe because a lot of stuff is happening in numpy arrays (or collections of numpy arrays) rather than dataframes?

**Comment by katie-lamb** [comment]
> What does "% of comparisons with no blocking" mean? Is that the fraction of records that are in the same block, and thus available to be considered as candidate matches? If so, those numbers are all tiny! Seems like the brute force annual blocking plus KNN is extremely effective, even if we're pretty generous with the value of K.
>
The number of comparisons post-blocking becomes `n_ferc1_records * k`, since we generate k candidates for each FERC1 record. The number of comparisons with no blocking is `n_ferc1_records * n_eia_records`. So this badly named header is the number of post-blocking comparisons divided by the number of no blocking comparisions, which is just `k/n_eia_records`.
> Why does it make sense to encode the categorical `fuel_type_code_pudl` column using TF-IDF rather than e.g. a one-hot encoder?
> 
It probably does make sense to use one hot encoding. TF-IDF would take into account the frequency of the fuel type across all the records, so "gas" would probably have lower weight than "nuclear". I could see this being helpful if two plant names are drastically different but they're both nuclear, then they're more likely to be a match than if two gas plants have very different names. But maybe I'm wrong in thinking the nuclear records would be more likely to be matched. I bet results are similar with one hot encoding, but I could put it into the embedder class to have as an option.
> Are you surprised that the equal weighting works so well? What does that mean about the information we're comparing?
I am a little surprised that it works this well. I would have thought we'd have to weight plant name and utility name higher. I'm hopeful that it means that a relatively simple matching model (something linear i.e. `splink`) should work well enough.

**Comment by katie-lamb** [comment]
I think this PR could benefit from a @zschira eye. 

I tried running this on all years of data and ran into memory issues when I tried to create a `SimilaritySearcher` object (the embedding matrices that get passed in are very large). I guess my first idea is to do a dimensionality reduction on the matrices (TF-IDF creates a ton of features). Since ideally this step gets us close to 100% recall of the training data, it would be a bummer to lower recall with a dimensionality reduction. My second idea is that maybe I shouldn't be passing around these large embedding matrices between objects and there needs to be a little redesign of this module. 

Since I can't create the `SimilaritySearcher` object, I haven't tried running FAISS similarity search on all years of data. But FAISS has a bunch of different time/memory optimization options and a GPU library version if things really are getting too big.

**Comment by katie-lamb** [comment]
> Where should blocking results be stored?

Using Dagster could be helpful here to store intermediate data products and stuff. Would integrate into PUDL better.

> Memory issues running on all years of data

With EPA CEMS Dagster reads in one year in at a time, then processes the years in parallel. Dagster has some logic for optimizing synchronization.

> What's up with that threading issue?

Zach will try and run this and see if the same issue occurs.

> How should we handle experiment tracking?

Dagster should be able to handle this. Create an asset for the candidate matches, model settings, match dataframe etc.

> Should the training data be package data?

Seems fine to just leave in the repo, if we need to track and change it later then we can figure out what to with it.

Next steps:
* A few changes then merge this PR in
* Zach starts thinking about Dagsterizing this repo
* Start thinking about putting performance metrics in the CI

**Review by zschira — APPROVED** [review]
@katie-lamb this PR seems good to me, expecially since much of the implementation will change with adding dagster.

As for the dagsterization and experiment/metric tracking, I've been thinking/researching and I think we can try to keep it pretty simple for now. We could use more full-featured tools like [mlflow](https://mlflow.org/docs/latest/index.html), but that seems like overkill for now, particularly given that we're only using unsupervised models and don't have to save a bunch of weights and stuff, and have pretty straightforward metrics we're looking at.

Given that, my proposed solution is:

1. Dagsterize through the blocking step with assets for the inputs, embedded dataframes, and candidate matches. Dagster will handle caching all these and hopefully solve some the memory/threading issues as well
2. Create a github action that will generate a csv (or alternative format) with metrics that we can cache to allow comparing between branches on PRs, a then use a [job-summary](https://github.com/marketplace/actions/github-actions-job-summary) to output a table with these metrics
